bin/put-multiarch.pl

```diff
@@ -31,39 +31,106 @@
 sub get_arch_p ($targetRef, $arch, $archRef) {
 	return $ua->get_manifest_p($archRef)->then(sub ($manifestData = undef) {
 		return unless $manifestData;
-		my ($digest, $manifest, $size) = ($manifestData->{digest}, $manifestData->{manifest}, $manifestData->{size});
+		my ($mediaType, $digest, $size, $manifest) = (
+			$manifestData->{mediaType},
+			$manifestData->{digest},
+			$manifestData->{size},
+			$manifestData->{manifest},
+		);
 
-		my $mediaType = $manifestData->{mediaType};
-		if ($mediaType eq Bashbrew::RegistryUserAgent::MEDIA_OCI_INDEX_V1 || $mediaType eq Bashbrew::RegistryUserAgent::MEDIA_MANIFEST_LIST) {
-			# jackpot -- if it's already a manifest list, the hard work is done!
-			return ($archRef, $manifest->{manifests});
-			# TODO we should validate the "platform" values here to make sure we're not violating the security boundary
+		my @manifests;
+		if (Bashbrew::RegistryUserAgent::is_media_image_list($mediaType)) {
+			push @manifests, @{ $manifest->{manifests} };
 		}
-		if ($mediaType eq Bashbrew::RegistryUserAgent::MEDIA_OCI_MANIFEST_V1 || $mediaType eq Bashbrew::RegistryUserAgent::MEDIA_MANIFEST_V1 || $mediaType eq Bashbrew::RegistryUserAgent::MEDIA_MANIFEST_V2) {
-			my $manifestListItem = {
+		elsif (Bashbrew::RegistryUserAgent::is_media_image_manifest($mediaType)) {
+			push @manifests, {
 				mediaType => $mediaType,
 				size => $size,
 				digest => $digest,
-				platform => {
-					arch_to_platform($arch),
-					($manifest->{'os.version'} ? ('os.version' => $manifest->{'os.version'}) : ()),
-				},
 			};
-			if ($manifestListItem->{platform}{os} eq 'windows' && !$manifestListItem->{platform}{'os.version'} && $mediaType eq Bashbrew::RegistryUserAgent::MEDIA_MANIFEST_V2) {
-				# if we're on Windows, we need to make an effort to fetch the "os.version" value from the config for the platform object
-				return $ua->get_blob_p($archRef->clone->digest($manifest->{config}{digest}))->then(sub ($config = undef) {
-					if ($config && $config->{'os.version'}) {
-						$manifestListItem->{platform}{'os.version'} = $config->{'os.version'};
-					}
-					return ($archRef, [ $manifestListItem ]);
-				});
+		}
+		else {
+			die "unknown mediaType '$mediaType' for '$archRef'";
+		}
+
+		# filter out objects we know we don't want (not a hashref, missing required fields)
+		@manifests = grep {
+			# https://specs.opencontainers.org/image-spec/descriptor/?v=v1.0.1
+			'HASH' eq ref($_) && $_->{mediaType} && $_->{digest} && $_->{size}
+		} @manifests;
+
+		# filter objects down to just fields we care about
+		@manifests = map {
+			# https://specs.opencontainers.org/image-spec/descriptor/?v=v1.0.1
+			{ %{ $_ }{qw{
+				mediaType
+				digest
+				size
+				annotations
```

Comment:
Hmmm, thinking about this, we should maybe have an explicit whitelist of acceptable annotations (eventually)?

(Perhaps better to handle that via PR review for https://github.com/docker-library/bashbrew/issues/51 images, and toolchain control otherwise so we don't have to change this file all the time when the expected annotations list changes over time?)

Reply:
:thinking: Probably an explicit allow list here (~in a configuration file?~ :speak_no_evil:). Does it make sense to also protect annotations from a misbehaving arch-specific server? :question:

Reply:
After thinking about it more, a misbehaving or malicious server is already serving potentially bad image content, so if we're going to protect against that properly we should probably do so with more proactive monitoring (not trying to make this script "smart" enough to prevent it heuristically somehow).